File: previous_test/2024_KAKAO_WINTER_INTERNSHIP.py
class Friend:
    def __init__(self, name):
        self.__name = name
        self.send_list = {}
        self.total_send = 0
        self.total_recieve = 0
        self.__expected_gift = 0

    def init_friends(self, friends):
        for friend in friends:
            if self.__name != friend:
                self.send_list[friend] = 0

    def send_gift(self, reciever: str):
        self.total_send += 1
        self.send_list[reciever] += 1

    def recieve_gift(self):
        self.total_recieve += 1

    def calculate_present_index(self) -> int:
        return self.total_send - self.total_recieve

    def get_expect_gift(self, given_dict) -> int:
        for k, v in self.send_list.items():
            reciever: Friend = given_dict[k]
            if v > reciever.send_list[self.__name]:
                self.__expected_gift += 1
            if v == reciever.send_list[self.__name]:
                if self.calculate_present_index() > reciever.calculate_present_index():
                    self.__expected_gift += 1

        return self.__expected_gift


def solution(friends, gifts):
    answer = 0
    given_dict = {}

    for friend in friends:
        new_friend = Friend(friend)
        new_friend.init_friends(friends=friends)
        given_dict[friend] = new_friend

    for gift in gifts:
        sender, reciever = gift.split()
        given_dict[sender].send_gift(reciever)
        given_dict[reciever].recieve_gift()

    for _, friend in given_dict.items():
        answer = max(answer, friend.get_expect_gift(given_dict))

    return answer


if __name__ == "__main__":
    result = solution(
        ["muzi", "ryan", "frodo", "neo"],
        [
            "muzi frodo",
            "muzi frodo",
            "ryan muzi",
            "ryan muzi",
            "ryan muzi",
            "frodo muzi",
            "frodo ryan",
            "neo muzi",
        ],
    )

    print(result)

File: previous_test/2024_KAKAO_WINTER_INTERNSHIP.java
import java.util.*;

class Friend {
    private final String name;
    private final Map<String, Integer> sendList = new HashMap<>();
    private int totalSend = 0;
    private int totalRecieve = 0;
    private int expectedGift = 0; 

    public Friend(String name, String[] friends) {
        this.name = name;
        this.initFirends(friends);
    }

    private void initFirends(String[] friends) {
        for (String friendName : friends) {
            if (!friendName.equals(this.name)) this.sendList.put(friendName, 0);
        }
    }

    public void sendGift(String reciever) {
        this.totalSend++;
        this.sendList.put(reciever, this.sendList.getOrDefault(reciever, 0) + 1);
    }

    public void recieveGift() {
        this.totalRecieve++;
    }

    public int getGiftIndex() {
        return this.totalSend - this.totalRecieve;
    }

    public int getExpectGift(Map<String, Friend> given) {
        for (String friendName : this.sendList.keySet()) {
            int count = this.sendList.get(friendName);
            Friend reciever = given.get(friendName);

            if (canExpected(count, reciever)) this.expectedGift++;
        }

        return this.expectedGift;
    }

    private boolean canExpected(int count, Friend reciever) {
        return count > reciever.sendList.get(this.name) ||
        (count == reciever.sendList.get(this.name) & this.getGiftIndex() > reciever.getGiftIndex());
    }
}

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;

        Map<String, Friend> given = new HashMap<>();

        for (String friend : friends) {
            Friend newFriend = new Friend(friend, friends);
            given.put(friend, newFriend);
        }

        for (String gift : gifts) {
            String[] splitGift = gift.split(" ");
            String sender = splitGift[0];
            String reciever = splitGift[1];

            given.get(sender).sendGift(reciever);
            given.get(reciever).recieveGift();
        }

        for (String friendName : given.keySet()) {
            answer = Math.max(answer, given.get(friendName).getExpectGift(given));
        }

        return answer;
    }
}